src/main/java/org/cloudfoundry/promregator/lite/discovery/SimpleTargetResolver.kt

```diff
@@ -0,0 +1,191 @@
+        private val cf: CloudFoundryConfig
+) {
+    suspend fun resolveTargets(configTargetsToMatch: List<Target>): List<AppResponse> = coroutineScope {
+        val configTargets = configTargetsToMatch.ifEmpty { listOf(Target()) }
+
+        val orgs = resolveOrgs(configTargets)
+        val domains = orgs.resolveDomains()
+        val spaces = orgs.resolveSpaces(configTargets, domains)
+
+        val apps = if (cfAccessor.isV3Enabled) {
+            val spaceSummaries = spaces.resolveSummaries()
+            spaces.resolveAppsV3(spaceSummaries)
+        } else {
+            spaces.resolveAppsV2()
+        }
+
+        return@coroutineScope apps.map { app ->
+            val target = configTargets.firstOrNull { t ->
+                app.space.org.api == t.api &&
+                        ((t.applicationName == null && t.applicationRegex == null) ||
+                        (t.applicationName == null && t.applicationRegex?.toRegex(IGNORE_CASE)?.matches(app.appName) == true) ||
+                        (t.applicationName == app.appName && t.applicationRegex == null))
+            }
+            app.copy(originalTarget = target)
+        }.filter { it.originalTarget != null && it.isScrappable }
+    }
+
+    private suspend fun resolveOrgs(configTargets: List<Target>): List<OrgResponse> = coroutineScope {
+        cf.api.keys.map { api ->
+            async {
+                cfAccessor.retrieveAllOrgIds(api).map {
+                    OrgResponse(api, it.metadata.id, it.entity.name)
+                }
+            }
+        }.awaitAll().flatten().filter { org ->
+            configTargets.any { target ->
+                org.api == target.api &&
+                        ((target.orgName == null && target.orgRegex == null) ||
+                        (target.orgName == null && target.orgRegex?.toRegex(IGNORE_CASE)?.matches(org.name) == true) ||
+                        (target.orgRegex == null && target.orgName == org.name))
+            }
+        }
+    }
+
+    private suspend fun List<OrgResponse>.resolveDomains() = coroutineScope {
+        return@coroutineScope this@resolveDomains.map { org ->
+            async {
+                cfAccessor.retrieveAllDomains(org.api, org.id)?.resources?.map { domainResource ->
+                    DomainResponse(org.id, domainResource.metadata.id, domainResource.entity.internal ?: false)
+                } ?: listOf()
+            }
+        }.awaitAll().flatten()
+    }
+
+    private suspend fun List<OrgResponse>.resolveSpaces(configTargets: List<Target>, domains: List<DomainResponse>) = coroutineScope {
+        return@coroutineScope this@resolveSpaces.map { org ->
+            async {
+                cfAccessor.retrieveSpaceIdsInOrg(org.api, org.id).map { space ->
+                    SpaceResponse(org, space.entity.name, space.metadata.id, domains.filter { it.orgId == org.id })
+                }
+            }
+        }.awaitAll().flatten().filter { space ->
```

Comment:
The `async` and `awaitAll` calls allow for the parallel processing that keeps this fast and easy to reason about. This is Kotlin coroutines being used. You'll notice the `suspend` keyword on the function. Coroutines yield the thread (suspending) during things like `await`ing so that is how the parallelism is managed.